src/types.rs

```diff
@@ -90,9 +90,191 @@ impl TypeTag {
     }
 }
 
+use itertools::FoldWhile::{Continue, Done};
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, derive_more::Display, Serialize, Deserialize)]
+/// Bounds on capabilities of a type.
+pub enum TypeBound {
+    /// The equality operation is valid on this type.
+    #[serde(rename = "E")]
+    Eq,
+    /// The type can be copied in the program.
+    #[serde(rename = "C")]
+    Copyable,
+}
+
+impl TypeBound {
+    /// Returns the smallest TypeTag containing both the receiver and argument.
+    /// (This will be one of the receiver or the argument.)
+    pub fn union(self, other: Self) -> Self {
+        if self.contains(other) {
+            self
+        } else {
+            debug_assert!(other.contains(self));
+            other
+        }
+    }
+
+    /// Report if this bound contains another.
+    pub fn contains(&self, other: TypeBound) -> bool {
+        use TypeBound::*;
+        match (self, other) {
+            (Copyable, Eq) => true,
+            (Eq, Copyable) => false,
+            _ => true,
+        }
+    }
+}
+
+/// Calculate the least upper bound for an iterator of bounds
+fn least_upper_bound(mut tags: impl Iterator<Item = Option<TypeBound>>) -> Option<TypeBound> {
+    tags.fold_while(Some(TypeBound::Eq), |acc, new| {
+        if let (Some(acc), Some(new)) = (acc, new) {
+            Continue(Some(acc.union(new)))
+        } else {
+            // if any type is unbounded, short-circuit
+            Done(None)
+        }
+    })
+    .into_inner()
+}
+
+use std::fmt::Write;
+
+use crate::{ops::AliasDecl, resource::PRELUDE, utils::display_list};
+use std::fmt::{self, Debug, Display};
+
+use self::type_param::TypeArg;
+
+//TODO remove
+type NewPrimType = primitive::PrimType;
+
+#[derive(Clone, PartialEq, Debug, Eq, derive_more::Display)]
+enum TypeEnum {
+    Prim(NewPrimType),
+    #[display(fmt = "Tuple({})", "DisplayRow(_0)")]
+    Tuple(Vec<Type>),
+    #[display(fmt = "Sum({})", "DisplayRow(_0)")]
+    Sum(Vec<Type>),
+}
+impl TypeEnum {
+    fn least_upper_bound(&self) -> Option<TypeBound> {
+        match self {
+            TypeEnum::Prim(p) => p.bound(),
+            TypeEnum::Tuple(ts) => least_upper_bound(ts.iter().map(Type::least_upper_bound)),
+            TypeEnum::Sum(ts) => least_upper_bound(ts.iter().map(Type::least_upper_bound)),
+        }
+    }
+}
+
+#[derive(
+    Clone, PartialEq, Debug, Eq, derive_more::Display, serde::Serialize, serde::Deserialize,
+)]
+#[display(fmt = "{}", "_0")]
+#[serde(into = "serialize::SerSimpleType", from = "serialize::SerSimpleType")]
+/// A HUGR type.
```

Comment:
This is a key thing in the public API, so it will need a more extensive explanation in the future.